applications/fishing-map/src/features/workspace/workspace.hook.ts

```diff
@@ -1,50 +1,52 @@
 import { useSelector } from 'react-redux'
 import { useCallback } from 'react'
-import { WorkspaceDataviewConfig } from '@globalfishingwatch/dataviews-client'
-import { selectDataviewsConfig } from 'routes/routes.selectors'
+import { UrlDataviewInstance } from 'types'
+import { selectDataviewInstances } from 'routes/routes.selectors'
 import { useLocationConnect } from 'routes/routes.hook'
-import { selectWorkspaceDataviewConfig } from './workspace.selectors'
+import { selectWorkspaceDataviewInstances } from './workspace.selectors'
 
-export const useDataviewsConfigConnect = () => {
-  const urlDataviewsConfig = useSelector(selectDataviewsConfig)
-  const workspaceDataviewsConfig = useSelector(selectWorkspaceDataviewConfig)
+export const useDataviewInstancesConnect = () => {
+  const urlDataviewInstances = useSelector(selectDataviewInstances)
+  const workspaceDataviewInstances = useSelector(selectWorkspaceDataviewInstances)
   const { dispatchQueryParams } = useLocationConnect()
 
-  const updateDataviewConfig = useCallback(
-    (dataviewConfig: Partial<WorkspaceDataviewConfig>) => {
-      const currentDataviewConfig = urlDataviewsConfig?.find(
-        (urlDataviewConfig) => urlDataviewConfig.id === dataviewConfig.id
+  const upsertDataviewInstance = useCallback(
```

Comment:
I'm torn about this. It's looks like a smart way to add or update an instance, but I also feel that you always know whether you're adding or updating (ie adding if loading track for search, updating if toggling visibility in HeatmapLayerPanel for instance).
I am worried we'll be doing by error `add` instead of `update` because somehow the id wrong.

- export two distinct explicit methods?
- or maybe throw an error if `id` is provided and dataviewInstance found in `urlDataviewInstances` (in which case the upsert should be responsible for generating an id when doing `add`, which might be a good idea, see my other comment)

Reply:
I'm not sure I got your point here, but some thoughts here:

About your first comment, I could create two separate methods to be able to use them explicitly when needed but then we'll need to duplicate logic to check first if it already exists in the URL.

Then, your worry will be even bigger as the logic of checking when you need to add or update is distributed on different components as this won't solve the id error because the id has always to be provided.